controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java

```diff
@@ -0,0 +1,198 @@
+                                        log.info("{}: successfully started bucket service bucket: {} ", BucketManager.this.serviceType, bucket);
+                                        bucketFuture.complete(null);
+                                    }
+
+                                    @Override
+                                    public void failed(State from, Throwable failure) {
+                                        super.failed(from, failure);
+                                        log.error("{}: Failed to start bucket: {} ", BucketManager.this.serviceType, bucket);
+                                        buckets.remove(bucket);
+                                        bucketFuture.completeExceptionally(failure);
+                                    }
+                                }, executor);
+
+                                bucketService.startAsync();
+                                return bucketService;
+                            } else {
+                                bucketFuture.complete(null);
+                                return y;
+                            }
+                        });
+
+                        return bucketFuture;
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        log.info("{}: Stop request received for bucket manager", serviceType);
+        Futures.allOf(buckets.values().stream().map(bucketService -> {
+            CompletableFuture<Void> bucketFuture = new CompletableFuture<>();
+            bucketService.addListener(new Listener() {
+                @Override
+                public void terminated(State from) {
+                    super.terminated(from);
+                    bucketFuture.complete(null);
+                }
+
+                @Override
+                public void failed(State from, Throwable failure) {
+                    super.failed(from, failure);
+                    bucketFuture.completeExceptionally(failure);
+                }
+            }, executor);
+            bucketService.stopAsync();
+
+            return bucketFuture;
+        }).collect(Collectors.toList()))
+                .whenComplete((r, e) -> {
+                    stopBucketOwnershipListener();
+                    if (e != null) {
+                        log.error("{}: bucket service shutdown failed with exception", serviceType, e);
+                        notifyFailed(e);
+                    } else {
+                        log.info("{}: bucket service stopped", serviceType);
+                        notifyStopped();
+                    }
+                });
+    }
+
+    public void notify(BucketNotification notification) {
+        switch (notification.getType()) {
+            case BucketAvailable:
+                RetryHelper.withIndefiniteRetriesAsync(() -> tryTakeOwnership(notification.getBucketId()),
```

Comment:
This behavior that "when a bucket becomes available take ownership of it" is here in the base class. But of the two implementation it really only applies to ZK because the in memory one is always the owner. Maybe this logic should move into the impl. (Just those few lines) The large chunk of code in the `thenCompose(isOwner -> {` callback on line 77 should remain in the base class in a proper method.

Reply:
The idea is that base class can have multiple implementation (in future we could have a store type other than zk and in memory). So the common logic of what to do when a bucket is available for ownership, independent of implementation, should reside in the base class. 
So here, if the `takeBucketOwnership` (in derived class) returns true then we know we have successfully taken ownership and we need to start `bucketService` instance for the said bucket. which is a shared piece of logic among implementations so it should be here itself.